### PR TITLE
feat(core): Add `ErrorCode` template to standardize conversion of user-defined error code enums to `std::error_code`.

### PR DESCRIPTION
**Pull request by LinZhihao-723**
# Description

## Problem Statement
In the current CLP core, we use a generic enum defined in `clp/ErrorCode.hpp` as our error code machism. We've been seeing problems of this design from the following perspectives:
- These error codes are generic, they don't provide adequate context about what exactly the error is about. They are used across multiple different components inside our codebase without detailed explanations. For example, `ErrorCode_Unsupported` is used in both reader interfaces to indicate unsupported operations, and it is also used in `WildcardToken` to report an unsupported wildcard token. This error code itself doesn't provide the upper caller with the context of the error.
- Different subcomponents may need their own error code enums. For example, the `ffi/ir_stream` component defines its own error enums `IRErrorCode` and `IRProtocalErrorCode`, instead of using the generic `ErrorCode_xxx`. There is no systematic way to organize these different error code enums across the code base.
- With the current design, we cannot integrate our error code into other error handling tools, such as `std_result` from `outcome`, or `std::expected` in the future if we migrate to C++23.

## Solution
In modern C++, the standard does provide a systematic and type-safe way of managing error codes: `<system_error>`. A detailed tutorial can be found in a cppcon talk [here][1].
The idea of this standard is simple. As module developers, you need to extend `std::error_code` to be compatible with a customized `std::error_category` and an enum of possible errors. The `std::error_category` will be used to identify the context of the errors with human-readable messages for each defined error. And for module users, they only need to compare the enum against the `std::error_code` instance to inspect errors.
Furthermore, the standard library also provide capabilities to define mappings from multiple error codes to one "error condition" using `std::error_condition`, providing better management for sources of errors.
By using this library, we should be able to solve all the problems mentioned in the previous section.

## Implementation
This PR implements `ErrorCode` as a generic error code template, which integrates `std::error_code` and can be extended to customized error enums.
To implement a new set of error codes, developers need to do the following steps:
1. Define a strongly typed enum of all possible errors, for example, `MyErrorCodeEnum`
2. Implement the specifications of the following methods from `ErrorCodeCategory<MyErrorCodeEnum>`:
    - `name()`: Defines the name of the error category
    - `message(MyErrorCodeEnum error_enum)`: Provides a translation from the error enum to an error message string
3. In `std` namespace, specialize the following template to be true:
```C++
template <>
struct is_error_code_enum<ErrorCode<MyErrorCodeEnum>> : std::true_type {};
```

After these three steps, `ErrorCode<MyErrorCodeEnum>` will be considered as a valid clp error code class. An instance of this class, initialized by an enum value from `MyErrorCodeEnum`, will be considered as a domain-specific error. This error can be implicitly converted into a `std::error_code`, providing compatibilities with `outcome::std_result` and user-level error inspections.
In addition to the template implementation, we also include a unit test to not only test the basic usage of the `ErrorCode` template, but also give an example of how to use the template to define a set of errors of a specific domain.
Last but not least, the current implementation should be extendable to add error condition supports in the future.

# Validation performed

1. Ensure all GitHub workflows passed.
2. Add unit tests to cover the basic usage of the template.
3. Tested with PR #482 to ensure this PR is compatible with the `std::error_code` handling introduced there.

[1]: https://www.youtube.com/watch?v=hNaLf8lYLDo

## Summary by CodeRabbit

- **New Features**
	- Introduced a structured error handling framework, allowing for the definition and management of custom error codes and categories.
	- Added comprehensive tests for the new error handling mechanisms to ensure reliability and correctness.

- **Bug Fixes**
	- Improved error categorization and messaging to enhance user experience during error handling.

### CODE REVIEW

**Review by davidlion — CHANGES_REQUESTED** [review]
Lets resolve some of the inconsistent naming:
* we switch between `err` and `error` throughout the code
* `errno` in `get_errno` and `error_num` in other places should be the same as far as I can tell
* `m_error` is a little ambiguous given almost everything is prefixed with "error" and its getter (currently `get_err_enum`) should match the name

I think it would be good if we add a test where we make a condition with a stdlib error_code and an error_code from your template to sanity check / ensure we don't break any functionality.

**Comment by LinZhihao-723** [comment]
Made some of the methods inline to the template declaration so that they can't be specialized

**Review by Bill-hbrhbr — CHANGES_REQUESTED** [review]
The major thing I'm concerned with is how to actually use this error-handling library beyond the examples provided in the unit tests.
From what I've seen after testing out this library myself:
1. The templates are implemented in the namespace `clp::error_handling`, so the overridden function implementations need to be in the same namespace as well. This is a bit counter-intuitive if you are working in a different namespace.
2. Ideally the ErrorCode enum class and the actual ErrorCode<ErrorCodeEnum> class should be placed in another header file so that the same error code system can be potentially used across different files. They could also be declared in a namespace that' could be entirely orthogonal to `clp::error_handling`. I don't have a clear clue on how that can be achieved. In the unit tests, these classes are defined in a global scope, and not in a header file. So a README that can guide library users would be very helpful.
3. Once the error code library works 100%, we can implement the same for the error conditions, and then figure out all possible scenarios of them comparing against each other.
4. we've discussed things like using small/harmless macros to help library users implementing their custom error codes. But this is not as important as figuring out how to use it properly first.

**Comment by kirkrodrigues** [comment]
On hold while we investigate modern ways of writing CMake scripts for libraries.

**Comment by LinZhihao-723** [comment]
> On hold while we investigate modern ways of writing CMake scripts for libraries.

After reading the CMakes of gsl and msgpack, I feel like we're doing header-only cmake correctly in this PR for libraries, right?

**Comment by kirkrodrigues** [comment]
> > On hold while we investigate modern ways of writing CMake scripts for libraries.
> 
> After reading the CMakes of gsl and msgpack, I feel like we're doing header-only cmake correctly in this PR for libraries, right?

I haven't read their header library code yet, but I know they both have fundamental bugs in their CMake scripts, so I wouldn't trust them.

File: components/core/src/clp/error_handling/ErrorCode.hpp [hunk 1]
```diff
@@ -0,0 +1,150 @@
...
+template <ErrorCodeEnumType ErrorCodeEnum>
+class ErrorCategory : public std::error_category {
+public:
+    // Methods implementing `std::error_category`
+    /**
+     * Gets the error category name.
+     * Note: A specialization must be explicitly implemented for each valid `ErrorCodeEnum`.
+     * @return The name of the error category.
+     */
+    [[nodiscard]] auto name() const noexcept -> char const* override;
```
[the remark this answers is not in the document]
Comment (LinZhihao-723) [reply components/core/src/clp/error_handling/ErrorCode.hpp:34]: This is wrong. These methods require the templated type to give a specialization. The class itself cannot be an abstracted class.

File: components/core/src/clp/error_handling/ErrorCode.hpp [hunk 2]
```diff
@@ -0,0 +1,150 @@
...
+        ErrorCodeEnum error_enum,
+        std::error_condition const& condition
+) const noexcept -> bool {
+    return std::error_category::default_error_condition(static_cast<int>(error_enum)) == condition;
+}
+
+template <typename ErrorCodeEnum>
+auto make_error_code(ErrorCode<ErrorCodeEnum> error) -> std::error_code {
+    return {error.get_error_num(), ErrorCode<ErrorCodeEnum>::get_category()};
+}
```
[the remark this answers is not in the document]
Comment (LinZhihao-723) [reply components/core/src/clp/error_handling/ErrorCode.hpp:134]: The class is designed to take a wrapper of the enum, but not the enum itself. The wrapper of the enum makes a general interface for adding customized error eums.

File: components/core/tests/test-error_handling.cpp [hunk 3]
```diff
@@ -0,0 +1,140 @@
...
+using BinaryErrorCode = ErrorCode<BinaryErrorCodeEnum>;
+using BinaryErrorCategory = ErrorCategory<BinaryErrorCodeEnum>;
+
+constexpr string_view cAlwaysSuccessErrorCategoryName{"Always Success Error Code"};
+constexpr string_view cBinaryTestErrorCategoryName{"Binary Error Code"};
+constexpr string_view cSuccessErrorMsg{"Success"};
+constexpr string_view cFailureErrorMsg{"Failure"};
+constexpr string_view cUnrecognizedErrorCode{"Unrecognized Error Code"};
+constexpr std::array cFailureConditions{std::errc::not_connected, std::errc::timed_out};
+constexpr std::array cNoneFailureConditions{std::errc::broken_pipe, std::errc::address_in_use};
```
[the remark this answers is not in the document]
Comment (LinZhihao-723) [reply components/core/tests/test-error_handling.cpp:38]: We're using C++20 and this piece of code compiled successfully for all our builds

File: components/core/tests/test-error_handling.cpp [hunk 4]
```diff
@@ -0,0 +1,140 @@
...
+        case BinaryErrorCodeEnum::Failure:
+            return std::any_of(
+                    cFailureConditions.cbegin(),
+                    cFailureConditions.cend(),
+                    [&](auto failure_condition) -> bool { return condition == failure_condition; }
+            );
+        default:
+            return false;
+    }
+}
```
[the remark this answers is not in the document]
Comment (LinZhihao-723) [reply components/core/tests/test-error_handling.cpp:91]: `AlwaysSuccessErrorCode` is an alias for `ErrorCode<AlwaysSuccessErrorCodeEnum>` (which is `clp::error_handling::ErrorCode` template), so `auto AlwaysSuccessErrorCategory::name() const noexcept -> char const*` should be equivalent to `auto clp::error_handling::ErrorCode<AlwaysSuccessErrorCodeEnum>::name() const noexcept -> char const*` unless I missed anything...

[the remark this answers is not in the document]
Comment (LinZhihao-723) [reply components/core/tests/test-error_handling.cpp:91]: Documentation: https://en.cppreference.com/w/cpp/language/template_specialization
```
namespace N
{
    template<class T> // primary template
    class X { /*...*/ };
    template<>        // specialization in same namespace
    class X<int> { /*...*/ };
 
    template<class T> // primary template
    class Y { /*...*/ };
    template<>        // forward declare specialization for double
    class Y<double>;
}
 
template<> // OK: specialization in same namespace
class N::Y<double> { /*...*/ };
```

**Review by Bill-hbrhbr — APPROVED** [review]

**Comment by Bill-hbrhbr** [comment]
Lgtm

**Review by davidlion — APPROVED** [review]

**Pull request merged**